### ISSUE TITLE
RFE: Module.customSections in the C API

### ISSUE THREAD

**Issue by FGasper** [wasmerio/wasmer#2285]
### Motivation

https://github.com/torch2424/as-bind/blob/5f277beb39a37371e84c0f2f8dd0c1e8fc7cf190/lib/asbind-instance/asbind-instance.js#L33

^^ I’d like to be able to implement this wrapper in languages other than JavaScript. I could do that readily enough if I had access to Module.customSections().

### Proposed solution

Mimic the JS interface in Rust and the C API.

### Alternatives

I don’t think there are any …

### Additional context

Once this is implemented it’ll be possible in languages like Perl to pass strings and arrays to WebAssembly via the language’s native syntax. That will be a boon for facilitating WASM adoption in use cases where, e.g., deduplication of validation logic between server-side and client-side logic is desired.

Thank you!

**Comment by syrusakbary** [comment]
Hi @FGasper,

Custom sections is actually already supported in Wasmer! via the `Module::custom_sections` method.
https://docs.rs/wasmer/1.0.2/wasmer/struct.Module.html#method.custom_sections

For a given `Module`, it will return an array with the contents of the different custom_sections for the provided name.

Let me know if that helps.
Closing the issue!

**Comment by FGasper** [comment]
@syrusakbary Oh cool!

Is that also available via the C API? I don’t see it in the header files that are installed on my system.

**Comment by syrusakbary** [comment]
Oh, I think we don't support via the C-API just yet, but it should be easy to add.
Would you like to contribute with a PR? :)

**Comment by FGasper** [comment]
I would _like_ to, yes, but my Rust skills aren’t where I’d like them to be. I’ll take a stab, though.

**Comment by FGasper** [comment]
Would you want to reopen this RFE, then? I’ll relabel to scope it specifically for the C API.

**Comment by syrusakbary** [comment]
Yup, I just reopened this PR and created an issue in the wasm-c-api repo!

**Comment by FGasper** [comment]
I see. So having this in Wasmer is, in the best-case scenario, contingent on its being in the external-standard wasm-c-api. That makes a lot of sense … I’ve subscribed to the other task and will chime in there if I can be of use.

Thank you!

**Comment by syrusakbary** [comment]
Ideally we will have it standardized within the Wasm C API.
We can start with a prototype C function and then propose that as the API!

**Comment by FGasper** [comment]
@syrusakbary Is the [currently-documented API](https://wasmerio.github.io/wasmer/c/) all deprecated? I note that stuff like `wasmer_instance_exports` is in a directory named `/deprecated/`.

**Comment by syrusakbary** [comment]
Yeah, I thought we updated the docs to reflect this changes, apologizes for the misdirection caused by that.

But yeah, the currently documented API is being deprecated in favor of the wasm-c-api.
Here are some fully working examples (check the `C/C++` tab).
https://docs.wasmer.io/integrations/examples/imports-and-exports

**Comment by Hywan** [comment]
The best way to get the C API documentation is still on docs.rs, just like any Wasmer crates, https://docs.rs/wasmer-c-api/1.0.2/wasmer_c_api/. You will find examples and everything you need, e.g. https://docs.rs/wasmer-c-api/1.0.2/wasmer_c_api/wasm_c_api/module/index.html.

I'll try to propose an API for the Wasm C API standard. If it's taking too long, I'll implement it in our `unstable` module. How does it sound to you @FGasper?

**Comment by FGasper** [comment]
@Hywan That sounds great—thank you!

There remain lots of links to Wasmer’s C API that either 404 or describe the legacy interface. Is that worth a separate issue filing?
